src/algebraic_geometry/projective_spectrum/structure_sheaf.lean

```diff
@@ -0,0 +1,477 @@
+      change localization.mk _ _ = localization.mk _ _,
+      simp only [num_zsmul, denom_zsmul],
+      convert congr_arg (λ z, m • z) h;
+      rw [zsmul_eq_mul, zsmul_eq_mul, show (m : at x) = localization.mk m 1, begin
+        induction m using int.induction_on with n ih n ih,
+        { erw localization.mk_zero _, refl, },
+        { push_cast,
+          erw [ih, ← localization.mk_one, localization.add_mk, one_mul, one_mul, add_comm, one_mul],
+          refl, },
+        { push_cast at ih ⊢,
+          erw [ih, ← localization.mk_one, sub_eq_add_neg, localization.neg_mk, localization.add_mk,
+            mul_one, one_mul, one_mul, add_comm, ← sub_eq_add_neg], },
+      end, localization.mk_mul, one_mul];
+      refl,
+    end) }
+
+instance homogeneous_localization.nat_scalar : has_scalar ℕ (homogeneous_localization x) :=
+{ smul := λ n z, (n : ℤ) • z }
+
+instance : has_neg (homogeneous_localization x) :=
+{ neg := quotient.map' has_neg.neg
+    (λ c1 c2 (h : localization.mk _ _ = localization.mk _ _), begin
+      change localization.mk _ _ = localization.mk _ _,
+      simp only [num_neg, denom_neg, ←localization.neg_mk],
+      exact congr_arg (λ c, -c) h
+    end) }
+
+instance : has_add (homogeneous_localization x) :=
+{ add := quotient.map₂' (+) (λ c1 c2 (h : localization.mk _ _ = localization.mk _ _)
+    c3 c4 (h' : localization.mk _ _ = localization.mk _ _), begin
+    change localization.mk _ _ = localization.mk _ _,
+    simp only [num_add, denom_add, ←localization.add_mk],
+    convert congr_arg2 (+) h h';
+    erw [localization.add_mk];
+    refl,
+  end) }
+
+instance : has_sub (homogeneous_localization x) :=
+{ sub := λ z1 z2, z1 + (-z2) }
+
+instance : has_mul (homogeneous_localization x) :=
+{ mul := quotient.map₂' (*) (λ c1 c2 (h : localization.mk _ _ = localization.mk _ _)
+    c3 c4 (h' : localization.mk _ _ = localization.mk _ _), begin
+    change localization.mk _ _ = localization.mk _ _,
+    simp only [num_mul, denom_mul],
+    convert congr_arg2 (*) h h';
+    erw [localization.mk_mul];
+    refl,
+  end) }
+
+instance : has_one (homogeneous_localization x) :=
+{ one := quotient.mk' 1 }
+
+instance : has_zero (homogeneous_localization x) :=
+{ zero := quotient.mk' 0 }
+
+lemma homogeneous_localization.zero_eq :
+  (0 : homogeneous_localization x) = quotient.mk' 0 :=
+rfl
+
+lemma homogeneous_localization.one_eq :
+  (1 : homogeneous_localization x) = quotient.mk' 1 :=
+rfl
+
+variable {x}
+lemma zero_val : (0 : homogeneous_localization x).val= 0 :=
+begin
+  rw [homogeneous_localization.zero_eq, homogeneous_localization.val, quotient.lift_on'_mk'],
+  change localization.mk _ _ = _,
+  convert localization.mk_zero _,
+end
+
+lemma one_val : (1 : homogeneous_localization x).val= 1 :=
+begin
+  rw [homogeneous_localization.one_eq, homogeneous_localization.val, quotient.lift_on'_mk'],
+  change localization.mk _ _ = _,
+  simp only [num_one, denom_one],
+  convert localization.mk_self _,
+  refl,
+end
```

Comment:
It's annoying how `irreducible` localization is, otherwise these proofs would all just be `rfl`

Reply:
Ah, this works!
```suggestion
lemma zero_val : (0 : homogeneous_localization x).val= 0 :=
localization.mk_zero _

lemma one_val : (1 : homogeneous_localization x).val= 1 :=
localization.mk_one
```

Reply:
Amazing